### PR TITLE
Allow unicode strings in the JSON schema

### PR DESCRIPTION
**Pull request by iamb**
### Summary
Allow JSON::Validator to validate schema with unicode strings.

### Motivation

I ran into this problem with a 3rd party json schema over which I have no control.

### References
This is the same issue described in #261 although I ran into it with an enum rather than uniqueItems.

The user of JSON::Validator does not have control over bytes vs. strings as given to data_checksum. When a schema comes from JSON, strings are decoded into strings of perl characters. The YAML modules decode to perl strings as well.

I had a similar fix to the one mentioned in that issue with a couple exceptions:

I did not copy the argument unless necessary because I wasn't sure if that was intentionally avoided in the original code.

I encode unconditionally the stringified scalar which is fed into the digest. This is intentional to avoid incorrectly accepted doubly-encoded UTF-8 as valid. That is demonstrated in the included tests.

I ran t/benchmark.t with n=200, n=500, and n=1000 against the previous code and this updated code and was not able to get any differences beyond noise.

### CODE REVIEW

File: lib/JSON/Validator/Util.pm [hunk 1]
```diff
@@ -24,7 +24,7 @@ sub E { JSON::Validator::Error->new(@_) }
 my $serializer = SEREAL_SUPPORT ? \&_sereal_encode : \&_yaml_dump;
 
 sub data_checksum {
-  return Mojo::Util::md5_sum(ref $_[0] ? $serializer->($_[0]) : defined $_[0] ? qq('$_[0]') : 'undef');
+  return Mojo::Util::md5_sum(ref $_[0] ? $serializer->($_[0]) : defined $_[0] ? do { utf8::encode(my $x = shift); qq('$x') } : 'undef');
```
Comment (jhthorsen) [inline lib/JSON/Validator/Util.pm:27]: Could you use encode from Mojo::Util instead? I also wonder what would happen if we run everything through $serializer...

File: t/spec/with-unicode-multibyte.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "object",
+  "properties": {
+    "foo": {
+      "enum": ["foo–bar"]
```
Comment (jhthorsen) [inline t/spec/with-unicode-multibyte.json:5]: Could you use a heart or something else that is easier to read than "–" ? Example: "json-validater ♥️ utf8"

File: t/unicode-multibyte.t [hunk 3]
```diff
@@ -0,0 +1,23 @@
+use lib '.';
+use t::Helper;
+use Mojo::File 'path';
+use Mojo::Util 'encode';
```
Comment (jhthorsen) [inline t/unicode-multibyte.t:4]: Can you use `qw(...)` instead of single quotes for consistency?

File: t/unicode-multibyte.t [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+use Test::More;
+
+my $builder = Test::More->builder;
+binmode($builder->output, ':encoding(UTF-8)');
+binmode($builder->failure_output, ':encoding(UTF-8)');
+binmode($builder->todo_output, ':encoding(UTF-8)');
```
Comment (jhthorsen) [inline t/unicode-multibyte.t:10]: I'm surprised that you have to do this, since I've never done that in any other unicode related tests. Can you explain?

**Review by jhthorsen — CHANGES_REQUESTED** [review]
I would be surprised if this PR doesn't get merged one way or another, but I would like to see some changes before that happens.

File: lib/JSON/Validator/Util.pm (hunk 1 above)
Comment (iamb) [reply lib/JSON/Validator/Util.pm:27]: I could, but chose specifically to use utf8::encode specifically. It shouldn't be possible for invalid codepoints (e.g., `\x{d800}`) to appear, but if they somehow did Mojo::Util (or Encode directly) would insert replacement characters. Different inputs (with invalid codepoints in the same place) would hash the same and compare the same. With utf8::encode, the perl representation is directly put into codepoints in the resulting string even if they are invalid. This way, different perl strings will always hash different.

If using Mojo::Util::encode

File: t/spec/with-unicode-multibyte.json (hunk 2 above)
Comment (iamb) [reply t/spec/with-unicode-multibyte.json:5]: Sure, will change. Happened to be the character I ran into.

File: t/unicode-multibyte.t (hunk 3 above)
Comment (iamb) [reply t/unicode-multibyte.t:4]: Will change

File: t/unicode-multibyte.t (hunk 4 above)
Comment (iamb) [reply t/unicode-multibyte.t:10]: This is just to avoid warnings when TAP is printing...for example:

````
t/unicode-multibyte.t .. 1/? Wide character in print at /opt/perl/perls/5.36.0/lib/5.36.0/Test2/Formatter/TAP.pm line 156.
Wide character in print at /opt/perl/perls/5.36.0/lib/5.36.0/Test2/Formatter/TAP.pm line 156.
t/unicode-multibyte.t .. ok
All tests successful.
Files=1, Tests=4,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.23 cusr  0.02 csys =  0.28 CPU)
Result: PASS
0
````

I'm not sure it's the best way either, but I don't usually deal with output unicode from tests. I can remove it if you prefer.

File: lib/JSON/Validator/Util.pm (hunk 1 above)
Comment (jhthorsen) [reply lib/JSON/Validator/Util.pm:27]: If that's the case then you have to add a(nother) test for that, or else I can pretty much guarantee you that I will change this later on and introduce a regression.

File: t/unicode-multibyte.t (hunk 4 above)
Comment (jhthorsen) [reply t/unicode-multibyte.t:10]: Sounds like validate_ok() should be fixed instead? If you think validate_ok() should be changed then please remove these three lines. If you want/can then please fix validate_ok() 😄

File: lib/JSON/Validator/Util.pm (hunk 1 above)
Comment (iamb) [reply lib/JSON/Validator/Util.pm:27]: gotcha, test added

File: t/unicode-multibyte.t (hunk 4 above)
Comment (iamb) [reply t/unicode-multibyte.t]: I'm pretty sure validate_ok() is correct here, the messages it's outputting are perl strings and it's up to the tap formatter to deal with output encoding. I need to research how to run tests better with expected unicode output (checking mojo for example, test names explicitly use "\uNNNN" rather than the characters)...I think that's on me. For now, I've removed from the tests.

File: lib/JSON/Validator/Util.pm (hunk 1 above)
Comment (iamb) [reply lib/JSON/Validator/Util.pm:27]: FWIW, I ran the benchmark test with n=1000 running this vs. just using serializer for everything. Using serializer was something like 35% slower in validating (just judging by repored time from the test).

Comment (jhthorsen) [reply lib/JSON/Validator/Util.pm:27]: Thanks for checking! I was expecting something like that, but I wasn't sure.

**Review by jhthorsen — APPROVED** [review]
